src/hotspot/share/opto/rangeinference.cpp

```diff
@@ -0,0 +1,536 @@
+    // Since all bits lower than the alignment are unset we don't need to
+    // align for the violations of ones anymore.
+    juint last_violation = W - 1 - count_leading_zeros(zero_violation);
+    T find_mask = std::numeric_limits<T>::max() << last_violation;
+    T either = lo | zeros;
+    T tmp = ~either & find_mask;
+    T alignment = tmp & (-tmp);
+    lo = (lo & -alignment) + alignment;
+    return lo | ones;
+  };
+
+  T new_lo = adjust_lo(lo, zeros, ones);
+  if (new_lo < lo) {
+    empty = true;
+    return true;
+  }
+
+  T new_hi = ~adjust_lo(~hi, ones, zeros);
+  if (new_hi > hi) {
+    empty = true;
+    return true;
+  }
+  bool progress = (new_lo != lo) || (new_hi != hi);
+  lo = new_lo;
+  hi = new_hi;
+  empty = lo > hi;
+  return progress;
+}
+
+template <class T>
+static bool adjust_bits_from_bounds(bool& empty, T& zeros, T& ones, T lo, T hi) {
+  static_assert(std::is_unsigned<T>::value, "");
+  T mismatch = lo ^ hi;
+  T match_mask = mismatch == 0 ? std::numeric_limits<T>::max()
+                               : ~(std::numeric_limits<T>::max() >> count_leading_zeros(mismatch));
+  T new_zeros = zeros | (match_mask &~ lo);
+  T new_ones = ones | (match_mask & lo);
+  bool progress = (new_zeros != zeros) || (new_ones != ones);
+  zeros = new_zeros;
+  ones = new_ones;
+  empty = ((zeros & ones) != 0);
+  return progress;
+}
+
+template <class T>
+static void normalize_constraints_simple(bool& empty, T& lo, T& hi, T& zeros, T& ones) {
+  adjust_bits_from_bounds(empty, zeros, ones, lo, hi);
+  if (empty) {
+    return;
+  }
+  while (true) {
+    bool progress = adjust_bounds_from_bits(empty, lo, hi, zeros, ones);
+    if (!progress || empty) {
+      return;
+    }
+    progress = adjust_bits_from_bounds(empty, zeros, ones, lo, hi);
+    if (!progress || empty) {
+      return;
+    }
+  }
+}
+
+template <class T, class U>
+void normalize_constraints(bool& empty, T& lo, T& hi, U& ulo, U& uhi, U& zeros, U& ones) {
+  static_assert(std::is_signed<T>::value, "");
+  static_assert(std::is_unsigned<U>::value, "");
+  static_assert(sizeof(T) == sizeof(U), "");
+
+  if (lo > hi || ulo > uhi || (zeros & ones) != 0) {
+    empty = true;
+    return;
+  }
+
+  if (T(ulo) > T(uhi)) {
+    if (T(uhi) < lo) {
+      uhi = std::numeric_limits<T>::max();
+    } else if (T(ulo) > hi) {
+      ulo = std::numeric_limits<T>::min();
+    }
+  }
+
+  if (T(ulo) <= T(uhi)) {
+    ulo = MAX2<T>(ulo, lo);
+    uhi = MIN2<T>(uhi, hi);
+    if (ulo > uhi) {
+      empty = true;
+      return;
+    }
+
+    normalize_constraints_simple(empty, ulo, uhi, zeros, ones);
+    lo = ulo;
+    hi = uhi;
+    return;
+  }
+
+  bool empty1 = false;
+  U lo1 = lo;
+  U hi1 = uhi;
+  U zeros1 = zeros;
+  U ones1 = ones;
+  normalize_constraints_simple(empty1, lo1, hi1, zeros1, ones1);
+
+  bool empty2 = false;
+  U lo2 = ulo;
+  U hi2 = hi;
+  U zeros2 = zeros;
+  U ones2 = ones;
+  normalize_constraints_simple(empty2, lo2, hi2, zeros2, ones2);
```

Comment:
At a quick glance of the code, it is not immediately clear why we need 2 ranges here. Can you add some comments, or maybe improve the naming from 1 and 2 to something more expressive?

Reply:
I have added explanations, basically the intersection of the signed and unsigned ranges is the union of 2 ranges, 1 in the negative range and the other in the positive one. We then process these separately and merge the results.